File: Data Structure/Array or Matrix/Medium/LET_no59_Spiral Matrix II.py
class Solution:
    def generateMatrix(self, n: int) -> List[List[int]]:
        res = [[0] * n for _ in range(n)]
        i = j = 0
        di, dj = 0, 1
        
        for cur in range(1, n * n + 1):
            res[i][j] = cur
            i2, j2 = i + di, j + dj
            if not (0 <= i2 < n) or not (0 <= j2 < n) or res[i2][j2] != 0:
                di, dj = dj, -di
            i += di
            j += dj
        
        return res
    

File: Data Structure/Array or Matrix/Medium/LET_no59_Spiral Matrix II.cpp
class Solution {
public:
    vector<vector<int>> generateMatrix(int n) {
        vector<vector<int>> res(n, vector<int>(n, 0));
        int val = 1, i = 0, j = 0, di = 0, dj = 1;
        while (val <= n * n) {
            res[i][j] = val;
            int i2 = (i + di + n) % n, j2 = (j + dj + n) % n;
            if (res[i2][j2] != 0) {
                int temp = dj;
                dj = -di, di = temp;
            }
            i += di;
            j += dj;
            val++;
        }

        return res;
    }
};